cpp/src/arrow/compute/exec/options.h

```diff
@@ -126,6 +126,25 @@ class ARROW_EXPORT SchemaSourceNodeOptions : public ExecNodeOptions {
   arrow::internal::Executor* io_executor;
 };
 
+class ARROW_EXPORT RecordBatchReaderSourceNodeOptions : public ExecNodeOptions {
+ public:
+  RecordBatchReaderSourceNodeOptions(std::shared_ptr<Schema> schema,
+                                     std::shared_ptr<RecordBatchReader> reader,
+                                     arrow::internal::Executor* io_executor = NULLPTR)
+      : schema(schema), reader(std::move(reader)), io_executor(io_executor) {}
+
+  /// \brief The schema of the record batches from the iterator
+  std::shared_ptr<Schema> schema;
```

Comment:
`RecordBatchReader` has a `schema()` method. We should use that instead of requiring it to be specified separately.